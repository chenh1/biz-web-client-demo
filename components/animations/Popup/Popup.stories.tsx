import type { AnimationBase } from '../../constants/types';
import React from 'react';
import { Popup } from './Popup';
import { Box } from '../../containers/Box/Box';
import { Heading } from '../../core/Heading/Heading';
import { Paragraph } from '../../core/Paragraph/Paragraph';
import { Button } from '../../core/Button/Button';

type Args = {
  args: AnimationBase
}

const Template = (args: AnimationBase) => (
  <Popup className={"max-w-xs"} {...args}>
    <Box className={"max-w-sm"} hasShadow isRounded gap="sm">
      <Heading type="h3">Heading text</Heading>
      <Paragraph>Example description</Paragraph>
      <Button label="CLICK ME" primary/>
    </Box>
  </Popup>
)

export default {
  component: Template
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const PopupAnimation: Args = { args: { disable: false } };

