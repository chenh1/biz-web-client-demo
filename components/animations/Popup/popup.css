@keyframes popup {
  0% {
    transform: scale(0, 0);
  }
  50% {
    transform: scale(1.08, 1.08);
  }
  75% {
    transform: scale(.95, .95);
  }
  88% {
    transform: scale(1.03, 1.03);
  }
  100% {
    transform: scale(1, 1);
  }
}


.popup {
  transform-origin: center;
  animation: popup .4s ease-in-out;
}

.freeze {
  animation-play-state: paused !important;
}