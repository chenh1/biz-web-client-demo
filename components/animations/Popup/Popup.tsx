'use client'

import type { AnimationBase } from '../../constants/types';
import React, { useState } from 'react';
import { useIntersectionObserver } from '../../hooks/useIntersectionObserver';
import "./popup.css"

export const Popup: React.FC<AnimationBase> = ({ className = "", disable, animateWhenInView = true, children }) => {
  const [ isInView, setIsInView ] = useState(false);

  const moduleRef = useIntersectionObserver({
    inViewCallback: () => {
      setIsInView(true)
    },
    leaveViewCallback: () => {
      setIsInView(false)
    }
  })

  return (
    <div ref={moduleRef} className={`relative ${!disable ? 'popup' : ''} ${!animateWhenInView || (animateWhenInView && isInView) ? '' : 'freeze'} ${className}`}>
      {children}
    </div>
  )
};