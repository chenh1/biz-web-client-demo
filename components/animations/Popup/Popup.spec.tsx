import { Popup } from "./Popup";
import { render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

const observe = jest.fn();
const unobserve = jest.fn();

// you can also pass the mock implementation
// to jest.fn as an argument
global.IntersectionObserver = jest.fn(function (callback, options) {
  return {
    observe,
    unobserve,
    root: null,
    rootMargin: '',
    thresholds: [],
    disconnect: jest.fn(),
    takeRecords: jest.fn(),
  };
});


it('pops up when in view', async () => {
  const { getByTestId, asFragment } = render(
    <Popup disable={false}>
      <p>Popup</p>
    </Popup>
  );

  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
});