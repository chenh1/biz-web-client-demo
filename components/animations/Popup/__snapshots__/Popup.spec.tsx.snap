// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pops up when in view 1`] = `
<DocumentFragment>
  <div
    class="relative popup freeze "
  >
    <p>
      Popup
    </p>
  </div>
</DocumentFragment>
`;
